packages/data-provider/src/utils/dictionarySchemaRelations.ts

```diff
@@ -47,3 +49,122 @@ export const getDictionarySchemaRelations = (
 		return acc;
 	}, {});
 };
+
+export interface TreeNode {
+	schemaName: string;
+	parentFieldName?: string;
+	fieldName?: string;
+	children?: TreeNode[];
+	parent?: TreeNode;
+}
+
+/**
+ * Function to find or create a node in the tree
+ * @param tree
+ * @param schemaName
+ * @param order
+ * @returns
+ */
+const findOrCreateNode = (tree: TreeNode[], schemaName: string, order: OrderType): TreeNode => {
+	let node = tree.find((n) => n.schemaName === schemaName);
+	if (!node) {
+		node = {
+			schemaName,
+			...(order === ORDER_TYPE.Values.desc ? { children: [] } : { parent: undefined }),
+		};
+		tree.push(node);
+	}
+	return node;
+};
+
+/**
+ * Finds a matching schema name within a nested object.
+ * Return true only if any matching schema name is found.
+ * @param treeNode
+ * @param schemaName
+ * @param type
+ * @returns
+ */
+const hasNestedNode = (treeNode: TreeNode, schemaName: string, type: SchemaRelationType): boolean => {
+	if (type === SCHEMA_RELATION_TYPE.Values.parent) {
+		if (!treeNode.parent) return false;
+		return hasNestedNode(treeNode.parent, schemaName, type);
+	} else {
+		if (!treeNode.children) return false;
+		return treeNode.children.some(
+			(node) =>
+				node.schemaName === schemaName ||
+				node.children?.some((innerNode) => hasNestedNode(innerNode, schemaName, type)),
+		);
+	}
+};
+
+/**
+ * Builds a hierarchy tree by linking the given schema to its parent schema based on foreign keys.
+ * This function recursively creates or finds nodes in the `tree` and establishes parent-child relationships
+ * between schemas using their foreign key mappings. It ensures that duplicate relationships are not created.
+ * @param schema The current schema definition to be added to the hierarchy tree
+ * @param tree The current tree of schema nodes, which gets updated with parent-child relationships as the hierarchy is built.
+ * @param order Order of the structure
+ * @returns
+ */
+const buildHierarchyTree = (schema: SchemaDefinition, tree: TreeNode[], order: OrderType) => {
+	// Create a node for the current schema
+	const node = findOrCreateNode(tree, schema.name, order);
+
+	schema.restrictions?.foreignKey?.forEach((foreignKey) => {
+		// Find the related schema by its name
+		const relatedSchema = findOrCreateNode(tree, foreignKey.schema, order);
+
+		// Use the first mapping for parent-child field relationship
+		const mapping = foreignKey.mappings[0];
```

Comment:
is this first item guaranteed?

Reply:
This logic is something we need to revisit. It works in all tested scenarios, however it's not fully guaranteed it works in more complex dictionaries. an issue was created #111